pkg/network/network.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2019 Sylabs, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package network
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/containernetworking/cni/libcni"
+	"github.com/golang/glog"
+	snetwork "github.com/sylabs/singularity/pkg/network"
+	k8s "k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2"
+)
+
+const (
+	// CNIBinDir is the default path to CNI plugin binaries.
+	CNIBinDir = "/opt/cni/bin"
+	// CNIConfDir is the default path to CNI network configuration files.
+	CNIConfDir = "/etc/cni/net.d"
+)
+
+// Manager contains network manager configuration and exposes
+// methods to bring up and down network interface.
+type Manager struct {
+	sync.RWMutex
+	defaultNetwork *libcni.NetworkConfigList
+	cniPath        *snetwork.CNIPath
+	podCidr        string
+}
+
+// PodNetworkConfig contains/defines POD network configuration.
+type PodNetworkConfig struct {
+	ID           string
+	Namespace    string
+	Name         string
+	NsPath       string
+	PortMappings []*k8s.PortMapping
+	Setup        *snetwork.Setup
+}
+
+// Init initialize CNI network manager.
+func (m *Manager) Init(cniPath *snetwork.CNIPath) error {
+	if m.cniPath != nil {
+		return nil
+	}
+	if cniPath == nil {
+		m.cniPath = &snetwork.CNIPath{
+			Plugin: CNIBinDir,
+			Conf:   CNIConfDir,
+		}
+	} else {
+		m.cniPath = cniPath
+	}
+	return m.setDefaultNetwork()
+}
+
+// checkInit updates CNI network configuration and does some
+// sanity checks.
+func (m *Manager) checkInit() error {
+	if err := m.setDefaultNetwork(); err != nil {
+		return err
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	ipRanges := false
+	for _, plugin := range m.defaultNetwork.Plugins {
+		if plugin.Network.Capabilities["ipRanges"] {
+			if m.podCidr == "" {
+				return fmt.Errorf("no PodCIDR set")
+			}
+			ipRanges = true
+			break
+		}
+	}
+	if !ipRanges && m.podCidr != "" {
+		glog.Infof("Reset pod CIDR, network plugin don't support it")
+		m.podCidr = ""
+	}
+	return nil
+}
+
+func (m *Manager) setDefaultNetwork() error {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.defaultNetwork != nil {
+		return nil
+	}
+	netConfList, err := snetwork.GetAllNetworkConfigList(m.cniPath)
+	if err != nil {
+		return fmt.Errorf("no CNI network configuration found in %s", m.cniPath.Conf)
```

Comment:
imo it is better to use `fmt.Errorf("could not get networks: %v", err)` here, so error and empty net list can be distinguished.